mailserver/cleaner.go

```diff
@@ -1,30 +1,91 @@
 package mailserver
 
 import (
+	"fmt"
+	"sync"
+	"time"
+
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
 	"github.com/syndtr/goleveldb/leveldb"
 	"github.com/syndtr/goleveldb/leveldb/iterator"
 	"github.com/syndtr/goleveldb/leveldb/util"
 )
 
-const batchSize = 1000
+const (
+	cleanerBatchSize = 1000
+	cleanerPeriod    = time.Hour
+)
+
+// cleaner removes old messages from a db.
+type cleaner struct {
+	sync.RWMutex
 
-// Cleaner removes old messages from a db
-type Cleaner struct {
 	db        dbImpl
 	batchSize int
+	retention time.Duration
+
+	period time.Duration
+	cancel chan struct{}
 }
 
-// NewCleanerWithDB returns a new Cleaner for db
-func NewCleanerWithDB(db dbImpl) *Cleaner {
-	return &Cleaner{
+// NewCleanerWithDB returns a new cleaner for db.
+func newCleanerWithDB(db dbImpl, retention time.Duration) *cleaner {
+	return &cleaner{
 		db:        db,
-		batchSize: batchSize,
+		retention: retention,
+
+		batchSize: cleanerBatchSize,
+		period:    cleanerPeriod,
+	}
+}
+
+// Start starts a loop that cleans up old messages.
+func (c *cleaner) Start() {
+	cancel := make(chan struct{})
+
+	c.Lock()
+	c.cancel = cancel
+	c.Unlock()
+
+	go c.schedule(c.period, cancel)
+}
+
+// Stops stops the cleaning loop.
+func (c *cleaner) Stop() {
+	c.Lock()
+	defer c.Unlock()
+
+	if c.cancel == nil {
+		return
+	}
+	close(c.cancel)
+	c.cancel = nil
+}
+
+func (c *cleaner) schedule(period time.Duration, cancel <-chan struct{}) {
+	t := time.NewTicker(period)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-t.C:
+			lower := uint32(0)
+			upper := uint32(time.Now().Add(-c.retention).Unix())
+			if _, err := c.Prune(lower, upper); err != nil {
+				log.Error("failed to prune data", "err", err)
+			}
+		case <-cancel:
+			return
+		}
 	}
 }
 
-// Prune removes messages sent between lower and upper timestamps and returns how many has been removed
-func (c *Cleaner) Prune(lower, upper uint32) (int, error) {
+// Prune removes messages sent between lower and upper timestamps
+// and returns how many have been removed.
+func (c *cleaner) Prune(lower, upper uint32) (int, error) {
```

Comment:
do we ever call it with different lower bound? maybe we can remove this parameter?
also, why not accepting `time.Time` here, why uint32?


something like 
```
PruneEntriesOlderThan(time.Now().Add(-c.retention))
```

Reply:
I can refactor it, I didn't really focus here on improving the interface. But yeah, it looks like the only call is here:
```go
lower := uint32(0)
upper := uint32(time.Now().Add(-c.retention).Unix())
if _, err := c.Prune(lower, upper); err != nil {
	log.Error("failed to prune data", "err", err)
}
```
so we can definitely change this method to accept a single param.

Reply:
yeah, and it makes sense to also make it accept `time.Time` too, that will avoid your `.Unix()` call when passing a param.